### ISSUE TITLE
Error messages in log

### ISSUE THREAD

**Issue by brianread108** [ddclient/ddclient#499]
Can someone explain what is happening here?

`FAILED:   updating bjsystems.myddns.me: unexpected status (15)
WARNING:  updating : nochg: No update required; unnecessary attempts to change to the current address are considered abusive
FAILED:   updating : unexpected status ()
FAILED:   updating : unexpected status (0)
FAILED:    was not updated because protocol <undefined> is not supported.

regards,
   [ddclient@mailserver.bjsystems.co.uk](mailto:ddclient@mailserver.bjsystems.co.uk) (version 3.10.0)`

Here is the ddclient.conf:

`daemon=300                                    # check every 300 seconds
syslog=yes                                     # log update msgs to syslog
pid=/var/run/ddclient.pid                      # record PID in file.
ssl=yes                                        # Set SSL

mail=admin
mail-failure=admin

use=web, web=checkip.amazonaws.com/, web-skip='IP Address:' 

server=dynupdate.no-ip.com, \
protocol=noip, \
login=xxxxxxxxxxxxxxxxxx, \
password=:xxxxxxxxxxxxxx \
bjsystems.myddns.me
`

**Comment by Musikolo** [comment]
Hi,

I'm getting a similar issue since I upgraded from `3.9.1` to `3.10.0` in my Arch Linux box. Getting the following error:

```
Jan 23 10:50:02 MyBox ddclient[658]: FAILED:   updating mybox.ddns.net: unexpected status (16)
Jan 23 10:50:02 MyBox ddclient[659]: WARNING:  updating : nochg: No update required; unnecessary attempts to change to the current address are considered abusive
Jan 23 10:50:02 MyBox ddclient[660]: FAILED:   updating : unexpected status ()
Jan 23 10:50:02 MyBox ddclient[661]: FAILED:   updating : unexpected status (0)
Jan 23 10:50:02 MyBox ddclient[662]: FAILED:    was not updated because protocol <undefined> is not supported.
```

Reverting back to `3.9.1` solves the issue.  My configuration has the following relevant configuration:
```
#server=dynupdate.no-ip.com
use=web, web=api.ipify.org
protocol=noip
```
 
 Is there anything  that needs to be changed from `3.10.0`?

**Comment by brianread108** [comment]
This happens sporadically, not obviously linked to any specific event. I have had it occurring on every refresh, but also it will go for days working fine.

If I restart ddclient, then I get this a few times:

`WARNING:  file /var/cache/ddclient/ddclient.cache, line 1: program version mismatch; ignoring /var/cache/ddclient/ddclient.cache
FAILED:   updating bjsystems.myddns.me: unexpected status (14)
WARNING:  updating : nochg: No update required; unnecessary attempts to change to the current address are considered abusive
FAILED:   updating : unexpected status ()
FAILED:   updating : unexpected status (0)
FAILED:    was not updated because protocol <undefined> is not supported`

**Comment by brianread108** [comment]
After a couple of restarts, I am now getting this every 3 minutes!!

`WARNING:  file /etc/ddclient/ddclient.conf: file /etc/ddclient/ddclient.conf should be owned only by ddclient or not be writable.
WARNING:  file /var/cache/ddclient/ddclient.cache: file /var/cache/ddclient/ddclient.cache should be owned only by ddclient or not be writable.
WARNING:  skipping update of bjsystems.myddns.me from <nothing> to 92.12.129.4.
WARNING:  last updated <never> but last attempt on Thu Jan 26 19:47:25 2023 failed.
WARNING:  Wait at least 5 minutes between update attempts.
regards,
   [ddclient@mailserver.bjsystems.co.uk](mailto:ddclient@mailserver.bjsystems.co.uk) (version 3.10.0)`

The permisssions on those files are this:

`
[root@mailserver ~]# ls -l /etc/ddclient/ddclient.conf 
-rw------- 1 ddclient ddclient 955 Jan 24 10:14 /etc/ddclient/ddclient.conf
[root@mailserver ~]# ls -l /var/cache/ddclient/ddclient.cache 
-rw------- 1 ddclient ddclient 232 Jan 26 19:51 /var/cache/ddclient/ddclient.cache
[root@mailserver ~]# 
`

**Comment by heeplr** [comment]
I'm getting those too for all noip domains. Debug output shows lots of warnings:

```
DEBUG:    
DEBUG:    nic_noip_update -------------------
DEBUG:    proxy    = <undefined>
DEBUG:    protocol = https
DEBUG:    server   = dynupdate.no-ip.com
DEBUG:    url      = nic/update?<redacted>
DEBUG:    ip ver   = 
FAILED:   updating myhost.hopto.org: unexpected status (15)
Use of uninitialized value $h in hash element at /usr/bin/ddclient line 4316.
Use of uninitialized value $_[0] in sprintf at /usr/bin/ddclient line 2160.
WARNING:  updating : nochg: No update required; unnecessary attempts to change to the current address are considered abusive
Use of uninitialized value $h in hash element at /usr/bin/ddclient line 4325.
Use of uninitialized value $h in hash element at /usr/bin/ddclient line 4326.
Use of uninitialized value $h in hash element at /usr/bin/ddclient line 4327.
Use of uninitialized value $h in hash element at /usr/bin/ddclient line 4316.
Use of uninitialized value $status in string eq at /usr/bin/ddclient line 4317.
Use of uninitialized value $status in exists at /usr/bin/ddclient line 4322.
Use of uninitialized value $status in pattern match (m//) at /usr/bin/ddclient line 4333.
Use of uninitialized value $status in numeric ne (!=) at /usr/bin/ddclient line 4345.
Use of uninitialized value $_[0] in sprintf at /usr/bin/ddclient line 2160.
FAILED:   updating : unexpected status ()
Use of uninitialized value $h in hash element at /usr/bin/ddclient line 4316.
Use of uninitialized value $_[0] in sprintf at /usr/bin/ddclient line 2160.
FAILED:   updating : unexpected status (0)
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
FAILED:    was not updated because protocol <undefined> is not supported.
DEBUG:    sleep 600
```

the config is:
```
protocol=noip
login=user@domain.com
password=redacted
myhost.hopto.org
```

Let me know if I can provide more info. Any hint on how to solve this highly appreciated.

**Comment by brianread108** [comment]
Definitely looks as though we have a problem specific to noip.  

Here is the structure internal to ddclient that describes the noip service:

  'noip' => {
        'updateable' => undef,
        'update'     => \&nic_noip_update,
        'examples'   => \&nic_noip_examples,
        'variables' => {
            'atime'                     => setv(T_NUMBER, 0, 1, 0,               undef),
            'custom'                    => setv(T_BOOL,   0, 1, 0,               undef),
            'host'                      => setv(T_STRING, 1, 1, '',              undef),
            'ip'                        => setv(T_IP,     0, 1, undef,           undef),
            'login'                     => setv(T_LOGIN,  1, 0, '',              undef),
            'max-interval'              => setv(T_DELAY,  0, 0, interval('25d'), 0),
            'min-error-interval'        => setv(T_DELAY,  0, 0, interval('5m'),  0),
            'min-interval'              => setv(T_DELAY,  0, 0, interval('30s'), 0),
            'mtime'                     => setv(T_NUMBER, 0, 1, 0,               undef),
            'password'                  => setv(T_PASSWD, 1, 0, '',              undef),
            'server'                    => setv(T_FQDNP,  1, 0, 'dynupdate.no-ip.com', undef),
            'static'                    => setv(T_BOOL,   0, 1, 0,               undef),
            'status'                    => setv(T_ANY,    0, 1, '',              undef),
            'warned-min-error-interval' => setv(T_ANY,    0, 1, 0,               undef),
            'warned-min-interval'       => setv(T_ANY,    0, 1, 0,               undef),
            'wtime'                     => setv(T_DELAY,  0, 1, 0,               interval('30s')),
        },

I am not familiar enough with the internals to know if this is the source of the problem.

The errors above occur in a SUB nic_noip_update.  Once again beyond my understanding of the program.

**Comment by darose** [comment]
This issue is definitely not specific to noip.  I'm having similar issues with dynu and OpenDNS:

```
# /usr/bin/ddclient -file /etc/ddclient/ddclient.conf         
CONNECT:  checkip.amazonaws.com                                                       
CONNECTED:  using SSL                                                                                                                                                        
SENDING:  GET / HTTP/1.1                                                              
SENDING:  Host: checkip.amazonaws.com                                                 
SENDING:  User-Agent: ddclient/3.10.0                                                 
SENDING:  Connection: close                                                           
SENDING:                                                                              
RECEIVE:  HTTP/1.1 200 OK                                                             
RECEIVE:  Date: Thu, 26 Jan 2023 20:35:00 GMT                          
RECEIVE:  Server: lighttpd/1.4.53                                                     
RECEIVE:  Content-Length: 14                                                          
RECEIVE:  Connection: Close                                                           
RECEIVE:                                                                                                                                                                     
RECEIVE:  <my.ip.address>                                                               
INFO:     forcing updating OpenDNSHome because no cached entry exists.
INFO:     forcing updating <my_hostname> because no cached entry exists.
INFO:     setting IP address to <my.ip.address> for <my_hostname>                                                                                                            
UPDATE:   updating <my_hostname>   
CONNECT:  api.dynu.com                                                                                                                                                       
CONNECTED:  using SSL               
SENDING:  GET /nic/update?system=dyndns&hostname=<my_hostname>&myip=<my.ip.address> HTTP/1.1
SENDING:  Host: api.dynu.com         
SENDING:  Authorization: Basic <my_auth_key>
SENDING:  User-Agent: ddclient/3.10.0                                                 
SENDING:  Connection: close                                                           
SENDING:                                                                              
RECEIVE:  HTTP/1.1 200 OK                                                             
RECEIVE:  Date: Thu, 26 Jan 2023 20:35:1 GMT              
RECEIVE:  Server: Dynu Web Server                                                     
RECEIVE:  X-Powered-By: Dynu Dynamic DNS Service                           
RECEIVE:  Content-Length: 5                                                           
RECEIVE:  Content-Type: text/html; charset=UTF-8                       
RECEIVE:                                                                              
RECEIVE:  nochg
WARNING:  updating <my_hostname>: nochg: No update required; unnecessary attempts to change to the current address are considered abusive
INFO:     setting IP address to <my.ip.address> for OpenDNSHome                                                                                                                
UPDATE:   updating OpenDNSHome                                                        
CONNECT:  updates.opendns.com                                                                                                                                                
CONNECTED:  using SSL
SENDING:  GET /nic/update?system=dyndns&hostname=OpenDNSHome&myip=<my.ip.address> HTTP/1.1
SENDING:  Host: updates.opendns.com
SENDING:  Authorization: Basic <my_auth_key>
SENDING:  User-Agent: ddclient/3.10.0
SENDING:  Connection: close
SENDING:  
RECEIVE:  HTTP/1.1 200 OK
RECEIVE:  date: Thu, 26 Jan 2023 20:35:02 GMT
RECEIVE:  server: opendns
RECEIVE:  vary: Accept-Encoding
RECEIVE:  content-type: text/html; charset=UTF-8
RECEIVE:  x-envoy-upstream-service-time: 49 
RECEIVE:  x-xss-protection: 1; mode=block
RECEIVE:  x-ingress-point: nyc
RECEIVE:  connection: close
RECEIVE:  transfer-encoding: chunked
RECEIVE:  
RECEIVE:  12
RECEIVE:  good <my.ip.address>
RECEIVE:  0
RECEIVE:  
FAILED:   updating OpenDNSHome: unexpected status (12)
Use of uninitialized value $h in hash element at /usr/bin/ddclient line 4101.
Use of uninitialized value $h in hash element at /usr/bin/ddclient line 4103.
Use of uninitialized value $h in hash element at /usr/bin/ddclient line 4104.
Use of uninitialized value $_[0] in sprintf at /usr/bin/ddclient line 2160.
SUCCESS:  updating : good: IP address set to <my.ip.address>
Use of uninitialized value $h in hash element at /usr/bin/ddclient line 4101.
Use of uninitialized value $_[0] in sprintf at /usr/bin/ddclient line 2160.
FAILED:   updating : unexpected status (0)
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
Use of uninitialized value in string ne at /usr/bin/ddclient line 1154.
FAILED:    was not updated because protocol <undefined> is not supported.

```

**Comment by Musikolo** [comment]
I found something interesting. It turns out that most options such as `use`,  `web`, `if` and many more has been divided into groups: the ones ending in `v4` and the ones ending in `v6`. So, if you have `use=web, web=example.org` you need to update it to the new syntax as `usev4=webv4, webv4=example.org`. After doing this, I don't get errors anymore, but I get the following warnings:

```
Jan 31 09:33:07 MyBox ddclient[21382]: WARNING:  file /var/cache/ddclient/ddclient.cache, line 1: program version mismatch; ignoring /var/cache/ddclient/ddclient.cache
Jan 31 09:33:07 MyBox ddclient[21383]: WARNING:  '' is not a valid IPv4 or IPv6 address
Jan 31 09:33:07 MyBox ddclient[21384]: WARNING:  found neither IPv4 nor IPv6 address
```

Not getting what I wanted, but I think it's a step forward. I recommend you to use `ddclient -h | less` to read the documentation and find all the new deprecations.

**Comment by Musikolo** [comment]
I found my issue, after setting `usev4=webv4, webv4=example.org` I noticed it was using SSL (port `443`) by default which isn't available in the endpoint I'm using. So, by disabling SSL altogether, it now works:  `usev4=webv4, webv4=example.org, ssl=no`.

I hope that helps!

**Comment by Musikolo** [comment]
Never mind, it still fails:
```
WARNING:  '' is not a valid IPv4 or IPv6 address
WARNING:  found neither IPv4 nor IPv6 address
FAILED:   updating myhost.ddns.net: unexpected status (16)
WARNING:  updating : nochg: No update required; unnecessary attempts to change to the current address are considered abusive
FAILED:   updating : unexpected status ()
FAILED:   updating : unexpected status (0)
FAILED:    was not updated because protocol <undefined> is not supported.
```

**Comment by heeplr** [comment]
TL;DR: Using the ```-curl``` parameter fixed it for me.

Here's what I found: I looked into the ```unexpected status (15)``` and found that parsing broke somehow between 3.9.x and 3.10.x:

ddclient expects a HTTP answer which separates headers and response payload by [an empty line](https://github.com/ddclient/ddclient/blob/master/ddclient.in#L4368) (switch to "results" state). So far so good. Then this empty line [should be followed](https://github.com/ddclient/ddclient/blob/master/ddclient.in#L4373) by a line in the format ```<status_string> <ip_address>```.

But all those debug dumps show, that before this line there is a number. (15 in my case with noip). Noip (and the other services) don't seem to send this 15. So I suspected ```geturl()``` is mangling the reply. Trying [curl](https://github.com/ddclient/ddclient/blob/master/ddclient.in#L2454) worked around the bug, suggesting it's down the ```fetch_via_socket_io()``` codepath.

I didn't look any further since I'm not too familiar to perl.

**Comment by Musikolo** [comment]
I looked into the code and I found `usev4=webv4` doesn't work in the the [get_ip()](https://github.com/ddclient/ddclient/blob/master/ddclient.in#L2851) function. It looks like it only supports the legacy values without `v4`or `v6` suffixes. So, I used `use=web` instead. For some reason, `webv4` attribute works well. Then, I got the same error as @heeplr  got, and tried using the property `curl=yes` as suggested. With this, I think it's working fine now. My final configuration is:

```
curl=yes
use=web, webv4=ip1.dynupdate.no-ip.com, ssl=no
```

I hope these issues are addressed for the next release.

Thanks.

**Comment by SuperSandro2000** [comment]
> I hope these issues are addressed for the next release.

If there is a PR for it, yes, otherwise it won't.

**Comment by Musikolo** [comment]
I've being doing some tests and I  found a configuration that has been working great for me for a few days. 

TL;DR: Using the `ssl=no` is now required for the `noip` protocol .

Although `use=web` and `web=someurl.example.org` options are deprecated, it's the only configuration that worked for me. Using the new `usev4=webv4`  and `webv4=someurl.example.org` options didn't work at all for me.

Also, when using `use=web` only, it doesn't use the correct built-in IP check URL, but `checkip.dyndns.org` instead. Next is the output I got when setting `verbose=yes`:
```
CONNECT:  checkip.dyndns.org
CONNECTED:  using HTTP
SENDING:  GET / HTTP/1.1
SENDING:  Host: checkip.dyndns.org
SENDING:  User-Agent: ddclient/3.10.0
SENDING:  Connection: close
SENDING:
RECEIVE:  HTTP/1.1 200 OK
RECEIVE:  Date: Sun, 05 Feb 2023 03:13:57 GMT
RECEIVE:  Content-Type: text/html
RECEIVE:  Content-Length: 106
RECEIVE:  Connection: close
RECEIVE:  Cache-Control: no-cache
RECEIVE:  Pragma: no-cache
RECEIVE:
RECEIVE:  <html><head><title>Current IP Check</title></head><body>Current IP Address: 93.184.216.34</body></html>
SUCCESS:  myhost.ddns.net: skipped: IP address was already set to 93.184.216.34.
```

As you can [see here](https://github.com/ddclient/ddclient/blob/master/ddclient.in#L117), the built-in web URL for `noip` is `ip1.dynupdate.no-ip.com`, but for some reason it ends up using  `checkip.dyndns.org` instead. This endpoint doesn't turns out to be as stable as `ip1.dynupdate.no-ip.com` is for me. So, to work around it, it needs to be explicitly set  in the `web` option.

So, my previous configuration (`use=web, web=ip1.dynupdate.no-ip.com`) now needs SSL explicitly disabled to work as expected: `use=web, web=ip1.dynupdate.no-ip.com, ssl=no`. I think this behavior changed from `3.9` to `3.10`, and the reason why the upgraded failed to me.

Finally, I didn't have any issues using the default HTTP client. So, I didn't have the need to enable curl (`curl=yes`) at all.

I hope this helps some of you.

**Comment by youduda** [comment]
I had a similar issue with dyndns2:
```
WARNING: skipping update of <domain> from <nothing> to <ip>.
FAILED: was not updated because protocol <undefined> is not supported.
FAILED: updating : unexpected status (0)
FAILED: updating : unexpected status ()
WARNING: updating : nochg: No update required; unnecessary attempts to change to the current address are considered abusive
FAILED: updating <domain>: unexpected status (13)
RECEIVE:
RECEIVE: 0
RECEIVE:
RECEIVE: nochg <ip>
RECEIVE: 13
RECEIVE:
RECEIVE: Vary: Accept-Encoding
RECEIVE: Vary: Accept-Encoding
RECEIVE: Connection: close
RECEIVE: Transfer-Encoding: chunked
RECEIVE: Content-Type: text/plain; charset=UTF-8
RECEIVE: Date: Tue, 28 Feb 2023 01:42:55 GMT
RECEIVE: Server: nginx
RECEIVE: HTTP/1.1 200 OK
```
Changing the implementation of the dyndns server to always set `Content-Length: xx` and therefore getting rid of `Transfer-Encoding: chunked` fixed it. This could be an issue with the HTTP 1.1 standard support.

**Comment by heeplr** [comment]
May I suggest to make curl the default (or even the only option, if that removes some code/complexity) as it seems to work for everyone, works around this bug and basically is probably available by default for every OS that runs ddclient.
